### PR TITLE
Add `M2M100TokenizerFast` (+ convert_slow_tokenizer implementation)

### PR DESCRIPTION
**Pull request by xenova**
# What does this PR do?

Adds fast tokenizer support for `M2M100Tokenizer`. I also added a `convert_slow_tokenizer` config to support generating a tokenizer.json file. For example, the generated tokenizer files for [facebook/m2m100_418M](https://huggingface.co/facebook/m2m100_418M) can be found [here](https://huggingface.co/Xenova/m2m100_418M/tree/main).

The fast tokenizer format is needed for transformers.js (see [issue](https://github.com/xenova/transformers.js/issues/235) and [PR](https://github.com/xenova/transformers.js/pull/250)). This may or may not be super relevant (considering the models are quite old), but there was a feature request in transformers.js, and I needed to write this code to get it working - so I thought I might as well share here.

I have ran the tokenizer on a [variety of test cases](https://github.com/xenova/transformers.js/blob/main/tests/generate_tests.py#L24-L46), and it passes each one. Additionally, it fixes a bug/inconsistency with the slow tokenizer (actually found in all sentencepiece tokenizers), where whitespace after special tokens is removed. To make the test cases past, I just basically hardcoded a [fix](https://github.com/xenova/transformers.js/blob/58425088a42931306304fa38b4cb50b901a2eefa/tests/generate_tests.py#L120-L125) for it (for now at least).

### Example conversion code
```python
from transformers import convert_slow_tokenizer, AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('Xenova/m2m100_418M', use_fast=False)
fast_tokenizer=convert_slow_tokenizer.convert_slow_tokenizer(tokenizer)
fast_tokenizer.save('tokenizer.json')
```

### Example usage code
```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('Xenova/m2m100_418M', use_fast=False)
fast_tokenizer = AutoTokenizer.from_pretrained('Xenova/m2m100_418M', use_fast=True)
assert tokenizer('Hello world').input_ids == fast_tokenizer('Hello world').input_ids
```

NOTE: To get facebook/m2m100_418M working, we need to remove the `"tokenizer_file": null,` line from [tokenizer_config.json](https://huggingface.co/facebook/m2m100_418M/blob/main/tokenizer_config.json)

Fixes # (issue)

### TODO
- [ ] Unit tests
- [x] Other config files + docs as done with `NllbTokenizerFast`
- [ ] Check that *all* special tokens are added (including `</s>` and similar tokens)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Feel free to ping me when you need a review!

**Comment by xenova** [comment]
@ArthurZucker Hey! 👋 I think this is ready for your review. The tokenizer is quite similar to NLLB, which was added [here](https://github.com/huggingface/transformers/pull/18126) and then updated by you [here](https://github.com/huggingface/transformers/pull/22313).

File: src/transformers/convert_slow_tokenizer.py [hunk 1]
```diff
@@ -772,6 +772,90 @@ def post_processor(self):
...
+        super().__init__(original_tokenizer)
+        original_tokenizer.vocab_file = self.original_vocab_file
+
+    def vocab(self, proto=None):
+        import json
+
+        with open(self.original_vocab_file) as fp:
+            vocab = json.load(fp)
+
+        vocab = list(sorted(vocab.items(), key=lambda x: x[1]))
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:789]: If `<s>` and `<unk>` etc are not part of the `original_vocab_file` then you need to add them here at the beginning.

File: src/transformers/models/m2m_100/tokenization_m2m_100_fast.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+        self.lang_code_to_token = {lang_code: f"__{lang_code}__" for lang_code in fairseq_language_code}
+
+        _additional_special_tokens = list(self.lang_code_to_token.values())
+        if additional_special_tokens is not None:
+            # Only add those special tokens if they are not already there.
+            _additional_special_tokens.extend(
+                [t for t in additional_special_tokens if t not in _additional_special_tokens]
+            )
+
+        self.add_special_tokens({"additional_special_tokens": _additional_special_tokens})
```
Comment (ArthurZucker) [inline src/transformers/models/m2m_100/tokenization_m2m_100_fast.py:182]: This should be adding the tokens to the list of special tokens, the problem is that in `fast` if the tokens are already part of the vocabulary they will not be added again. For Llama we used: 
```python 
            tokenizer.add_special_tokens(
                [
                    AddedToken("<unk>"),
                    AddedToken("<s>"),
                    AddedToken("</s>"),
                ]
            )
```
in the `convert_slow_tokenizer.py` script.

Did you convert the tokenizer slow using `from_slow=True`? 
(I am not really suprised by this bug)

File: src/transformers/models/m2m_100/tokenization_m2m_100_fast.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    @property
+    def src_lang(self) -> str:
+        return self._src_lang
+
+    @src_lang.setter
+    def src_lang(self, new_src_lang: str) -> None:
+        self._src_lang = new_src_lang
+        self.set_src_lang_special_tokens(self._src_lang)
+
+    def build_inputs_with_special_tokens(
```
Comment (ArthurZucker) [inline src/transformers/models/m2m_100/tokenization_m2m_100_fast.py:204]: Missing copied from statements here and wherever you copied from!

**Review by ArthurZucker — COMMENTED** [review]
Hey! Looks good already, could you give me more details on how you converted the model? The tests should also be updated I think (apply integration tests to both tokenizers!)

**Comment by xenova** [comment]
> Looks good already, could you give me more details on how you converted the model?

I just used [Optimum](https://github.com/huggingface/optimum) to convert the model to ONNX, but this shouldn't matter for this PR (which only concerns the tokenizer).

File: src/transformers/models/m2m_100/tokenization_m2m_100_fast.py (hunk 2 above)
Comment (xenova) [reply src/transformers/models/m2m_100/tokenization_m2m_100_fast.py:182]: > Did you convert the tokenizer slow using `from_slow=True`?

I didn't add any other parameters other than the ones in the original comment 😅 ... Is this needed, considering there is no fast tokenizer to convert from?

**Comment by ArthurZucker** [comment]
Sorry by model I meant the tokenizer model (the backend sentencepiece model if you will). I am trying to understand why you had to manually add the added tokens, and the `from_slow` is to initialize a fast tokenizer from a slow tokenizer using the conversion method!

**Comment by xenova** [comment]
@ArthurZucker I added the "copied from" annotations as well as added the special tokens (like they are done for Llama). On the latter point, the tokenizer includes "<madeupwordX>" tokens (for some reason) - should these be classified as special tokens?

File: src/transformers/models/m2m_100/tokenization_m2m_100_fast.py [hunk 4]
```diff
@@ -201,14 +203,15 @@ def src_lang(self, new_src_lang: str) -> None:
         self._src_lang = new_src_lang
         self.set_src_lang_special_tokens(self._src_lang)
 
+    # Copied from transformers.models.nllb.tokenization_nllb_fast.NllbTokenizerFast.build_inputs_with_special_tokens
```
Comment (ArthurZucker) [inline src/transformers/models/m2m_100/tokenization_m2m_100_fast.py:206]: You can add 
```suggestion
    # Copied from transformers.models.nllb.tokenization_nllb_fast.NllbTokenizerFast.build_inputs_with_special_tokens with NLLB->M2M100
```
😉

**Comment by ArthurZucker** [comment]
Regarding the madeup words, it depends, would just say let's follow what's done for slow!

**Comment by LysandreJik** [comment]
Should this be reopened @xenova ? :)

**Comment by xenova** [comment]
I marked it as a draft since I have been quite busy on some more important things, with the idea that I would return to it eventually once I had more time 😅. It's basically just missing some unit tests and example tokenizer files.

**Pull request closed without merging**